matrix_sdk_base/src/client.rs

```diff
@@ -1406,41 +1612,155 @@ impl BaseClient {
         };
 
         match event {
-            RoomEvent::RoomMember(mem) => event_emitter.on_room_member(room, &mem).await,
-            RoomEvent::RoomName(name) => event_emitter.on_room_name(room, &name).await,
-            RoomEvent::RoomCanonicalAlias(canonical) => {
-                event_emitter
-                    .on_room_canonical_alias(room, &canonical)
-                    .await
-            }
-            RoomEvent::RoomAliases(aliases) => event_emitter.on_room_aliases(room, &aliases).await,
-            RoomEvent::RoomAvatar(avatar) => event_emitter.on_room_avatar(room, &avatar).await,
-            RoomEvent::RoomMessage(msg) => event_emitter.on_room_message(room, &msg).await,
-            RoomEvent::RoomMessageFeedback(msg_feedback) => {
-                event_emitter
-                    .on_room_message_feedback(room, &msg_feedback)
-                    .await
-            }
-            RoomEvent::RoomRedaction(redaction) => {
-                event_emitter.on_room_redaction(room, &redaction).await
-            }
-            RoomEvent::RoomPowerLevels(power) => {
-                event_emitter.on_room_power_levels(room, &power).await
-            }
-            RoomEvent::RoomTombstone(tomb) => event_emitter.on_room_tombstone(room, &tomb).await,
-            RoomEvent::CustomRoom(custom) => {
-                event_emitter
-                    .on_unrecognized_event(room, &CustomOrRawEvent::CustomRoom(custom))
-                    .await
+            AnyRoomEventStub::StateEvent(event) => match &event.content {
+                AnyStateEventContent::RoomMember(mem) => {
+                    let prev = event
+                        .prev_content
+                        .as_ref()
+                        .map(|p| {
+                            if let AnyStateEventContent::RoomMember(m) = p {
+                                Some(m.clone())
+                            } else {
+                                None
+                            }
+                        })
+                        .flatten();
+                    event_emitter
+                        .on_room_member(room, &map_state_stub(event, mem.clone(), prev))
+                        .await
+                }
```

Comment:
This allows the EventEmitter to receive events with a specific content field as apposed to `StateEvent<AnyStateEventContent>`. It does add a lot of code but there was talk of adding a `map_content` method to the event structs.